BraveWallet/Crypto/Stores/SwapTokenStore.swift

```diff
@@ -32,48 +33,384 @@ public class SwapTokenStore: ObservableObject {
+          to: fromToken.contractAddress,
+          value: "0x0",
+          data: data
+        )
+      self?.transactionController.addUnapprovedTransaction(
+          txData,
+          from: accountInfo.address,
+          completion: { success, txMetaId, error in
+            // should be observed
+          }
+      )
+    }
+  }
+  
+  private func gweiStringToDecimal(_ value: String) -> BDouble? {
+    guard !value.isEmpty, let bv = BDouble(value) else {
+      return nil
+    }
+    return bv / (BDouble(10) ** 9)
+  }
+  
+  private func checkBalanceShowError(swapResponse: BraveWallet.SwapResponse) {
+    guard
+      let accountInfo = accountInfo,
+      let sellAmountValue = BDouble(sellAmount),
+      let gasLimit = gweiStringToDecimal(swapResponse.estimatedGas),
+      let gasPrice = gweiStringToDecimal(swapResponse.gasPrice),
+      let fromToken = selectedFromToken
+    else { return }
+    
+    // Check if balance is insufficient
+    if sellAmountValue > (selectedFromTokenBalance ?? 0) {
+      state = .error(Strings.Wallet.InsufficientBalance)
+    }
+    
+    // Get ETH balance for this account because gas can only be paid in ETH
+    rpcController.balance(accountInfo.address) { [weak self] success, balance in
+      guard let self = self else { return }
+      if success {
+        let fee = gasLimit * gasPrice
+        let balanceFormatter = WeiFormatter(decimalFormatStyle: .balance)
+        let currentBalance = BDouble(balanceFormatter.decimalString(for: balance.removingHexPrefix, radix: .hex, decimals: 18) ?? "") ?? 0
+        if fromToken.isETH {
+          if currentBalance < fee + sellAmountValue {
+            self.state = .error(Strings.Wallet.InsufficientFundsForGas)
+            return
+          }
+        } else {
+          if currentBalance < fee {
+            self.state = .error(Strings.Wallet.InsufficientFundsForGas)
+            return
+          }
+        }
+      }
+      
+      self.state = .swap
+      // check for ERC20 token allowance
+      if fromToken.isErc20 {
+        self.checkAllowance(
+          contractAddress: accountInfo.address,
+          spenderAddress: swapResponse.allowanceTarget,
+          amountToSend: sellAmountValue,
+          fromToken: fromToken
+        )
+      }
+    }
+  }
+  
+  private func checkAllowance(
+    contractAddress: String,
+    spenderAddress: String,
+    amountToSend: BDouble,
+    fromToken: BraveWallet.ERCToken
+  ) {
+    rpcController.erc20TokenAllowance(
+      contractAddress,
+      ownerAddress: contractAddress,
+      spenderAddress: spenderAddress
+    ) { [weak self] success, allowance in
+      let weiFormatter = WeiFormatter(decimalFormatStyle: .decimals(precision: 18))
+      let allowanceValue = BDouble(weiFormatter.decimalString(for: allowance.removingHexPrefix, radix: .hex, decimals: Int(fromToken.decimals)) ?? "") ?? 0
+      guard success, amountToSend > allowanceValue else { return } // no need to activate allowance
```

Comment:
Fallback to 0 on failure to convert means this guard will always pass, is that expected?

Reply:
I think so. 0 should be the min number. if we pass it here. worst case is failed at makeErc20ApproveData which will keep the same state and failed at addUnapprovedTransaction who's error will be handled separately